File: source/libear_a/Execution.cc
/*  Copyright (C) 2012-2017 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "libear_a/Execution.h"

#include <functional>

#include "libear_a/Array.h"

namespace {

    class ExecutionSerializer : public ear::Serializable {
    public:
        using Estimator = std::function<size_t()>;
        using Copier = std::function<char const **(char const **, char const **)>;

    public:
        ExecutionSerializer(Estimator const &estimator, Copier const &copier) noexcept
                : estimator_(estimator), copier_(copier) {}

        size_t estimate() const noexcept override {
            return estimator_();
        }

        const char **copy(char const **begin, char const **end) const noexcept override {
            return copier_(begin, end);
        }

    private:
        Estimator const &estimator_;
        Copier const &copier_;
    };

    // TODO: use result type for this method (to do proper error handling).
    int forward(ear::Serializable const &session,
                ear::Serializable const &execution,
                std::function<int(const char *, const char **)> const &function) noexcept {
        size_t const size = session.estimate() + execution.estimate();
        char const *dst[size];
        char const **it = dst;
        char const **const end = it + size;

        it = session.copy(it, end);
        it = execution.copy(it, end);

        return function(dst[0], dst);
    }

    int forward(ear::DynamicLinker const &linker,
                ear::Serializable const &session,
                ear::Serializable const &execution,
                char const **envp) noexcept {
        auto fp = linker.execve();
        return ::forward(session,
                       execution,
                       [envp, &fp](auto cmd, auto args) {
                           return fp(cmd,
                                     const_cast<char *const *>(args),
                                     const_cast<char *const *>(envp));
                       });
    }

}


namespace ear {

    int Execution::apply(DynamicLinker const &linker, State const *state) noexcept {
        return (state == nullptr)
               ? this->apply(linker)
               : this->apply(linker, LibrarySessionSerializer(state->get_input()));
    }

    Execve::Execve(const char *path, char *const *argv, char *const *envp) noexcept
            : path_(path)
            , argv_(const_cast<const char **>(argv))
            , envp_(const_cast<const char **>(envp))
    { }

    int Execve::apply(DynamicLinker const &linker) noexcept {
        auto fp = linker.execve();
        return fp(path_, const_cast<char *const *>(argv_), const_cast<char *const *>(envp_));
    }

    int Execve::apply(DynamicLinker const &linker, Serializable const &session) noexcept {
        ExecutionSerializer execution(
                [this]() {
                    return ::ear::array::length(argv_) + 2;
                },
                [this](auto begin, auto end) {
                    const char **argv_begin = argv_;
                    const char **argv_end = argv_begin + ::ear::array::length(argv_);

                    auto it = begin;
                    *it++ = command_separator;
                    return ::ear::array::copy(argv_begin, argv_end, it, end);
                }
        );
        return forward(linker, session, execution, envp_);
    }

    Execvpe::Execvpe(const char *file, char *const *argv, char *const *envp) noexcept
            : file_(file)
            , argv_(const_cast<char const **>(argv))
            , envp_(const_cast<char const **>(envp))
    { }

    int Execvpe::apply(DynamicLinker const &linker) noexcept {
        auto fp = linker.execvpe();
        return fp(file_, const_cast<char *const *>(argv_), const_cast<char *const *>(envp_));
    }

    int Execvpe::apply(DynamicLinker const &linker, Serializable const &session) noexcept {
        ExecutionSerializer execution(
                [this]() {
                    return ::ear::array::length(argv_) + 4;
                },
                [this](auto begin, auto end) {
                    const char **argv_begin = argv_;
                    const char **argv_end = argv_begin + ::ear::array::length(argv_);

                    auto it = begin;
                    *it++ = file_flag;
                    *it++ = file_;
                    *it++ = command_separator;
                    return ::ear::array::copy(argv_begin, argv_end, it, end);
                }
        );
        return forward(linker, session, execution, envp_);
    }

    ExecvP::ExecvP(const char *file, const char *search_path, char *const *argv, char *const envp[]) noexcept
            : file_(file)
            , search_path_(search_path)
            , argv_(const_cast<char const **>(argv))
            , envp_(const_cast<char const **>(envp))
    { }

    int ExecvP::apply(DynamicLinker const &linker) noexcept {
        auto fp = linker.execvP();
        return fp(file_, search_path_, const_cast<char *const *>(argv_));
    }

    int ExecvP::apply(DynamicLinker const &linker, Serializable const &session) noexcept {
        ExecutionSerializer execution(
                [this]() {
                    return ::ear::array::length(argv_) + 6;
                },
                [this](auto begin, auto end) {
                    const char **argv_begin = argv_;
                    const char **argv_end = argv_begin + ::ear::array::length(argv_);

                    auto it = begin;
                    *it++ = file_flag;
                    *it++ = file_;
                    *it++ = search_flag;
                    *it++ = search_path_;
                    *it++ = command_separator;
                    return ::ear::array::copy(argv_begin, argv_end, it, end);
                }
        );
        return forward(linker, session, execution, envp_);
    }

    Spawn::Spawn(pid_t *pid, const char *path, const posix_spawn_file_actions_t *file_actions,
                 const posix_spawnattr_t *attrp, char *const *argv, char *const *envp) noexcept
            : pid_(pid)
            , path_(path)
            , file_actions_(file_actions)
            , attrp_(attrp)
            , argv_(const_cast<const char **>(argv))
            , envp_(const_cast<const char **>(envp))
    { }

    int Spawn::apply(DynamicLinker const &linker) noexcept {
        auto fp = linker.posix_spawn();
        return fp(pid_, path_, file_actions_, attrp_, const_cast<char *const *>(argv_), const_cast<char *const *>(envp_));
    }

    int Spawn::apply(DynamicLinker const &linker, Serializable const &session) noexcept {
        ExecutionSerializer execution(
                [this]() {
                    return ::ear::array::length(argv_) + 2;
                },
                [this](auto begin, auto end) {
                    const char **argv_begin = argv_;
                    const char **argv_end = argv_begin + ::ear::array::length(argv_);

                    auto it = begin;
                    *it++ = command_separator;
                    return ::ear::array::copy(argv_begin, argv_end, it, end);
                }
        );
        auto fp = linker.posix_spawn();
        return ::forward(session,
                         execution,
                         [this, &fp](auto cmd, auto args) {
                             return fp(pid_,
                                       cmd,
                                       file_actions_,
                                       attrp_,
                                       const_cast<char *const *>(args),
                                       const_cast<char *const *>(envp_));
                         });
    }

    Spawnp::Spawnp(pid_t *pid, const char *file, const posix_spawn_file_actions_t *file_actions,
                   const posix_spawnattr_t *attrp, char *const *argv, char *const *envp) noexcept
            : pid_(pid)
            , file_(file)
            , file_actions_(file_actions)
            , attrp_(attrp)
            , argv_(const_cast<const char **>(argv))
            , envp_(const_cast<const char **>(envp))
    { }

    int Spawnp::apply(DynamicLinker const &linker) noexcept {
        auto fp = linker.posix_spawnp();
        return fp(pid_, file_, file_actions_, attrp_, const_cast<char *const *>(argv_), const_cast<char *const *>(envp_));
    }

    int Spawnp::apply(DynamicLinker const &linker, Serializable const &session) noexcept {
        ExecutionSerializer execution(
                [this]() {
                    return ::ear::array::length(argv_) + 4;
                },
                [this](auto begin, auto end) {
                    const char **argv_begin = argv_;
                    const char **argv_end = argv_begin + ::ear::array::length(argv_);

                    auto it = begin;
                    *it++ = file_flag;
                    *it++ = file_;
                    *it++ = command_separator;
                    return ::ear::array::copy(argv_begin, argv_end, it, end);
                }
        );
        auto fp = linker.posix_spawn();
        return ::forward(session,
                         execution,
                         [this, &fp](auto cmd, auto args) {
                             return fp(pid_,
                                       cmd,
                                       file_actions_,
                                       attrp_,
                                       const_cast<char *const *>(args),
                                       const_cast<char *const *>(envp_));
                         });
    }
}

File: rust/src/compilation/execution.rs
/*  Copyright (C) 2012-2018 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
use std::path;

use crate::{ErrorKind, Result};
use super::compiler;
use super::flags;
use super::pass;
use super::database;

#[derive(Default)]
struct CompilerExecution {
    directory: path::PathBuf,
    compiler: path::PathBuf,
    pass: pass::CompilerPass,
    flags: Vec<String>,
    inputs: Vec<path::PathBuf>,
    output: Option<path::PathBuf>,
}

impl CompilerExecution {
    fn new(working_dir: &str, compiler: &str) -> Self {
        let mut result: Self = Default::default();
        result.directory = path::PathBuf::from(working_dir);
        result.compiler = path::PathBuf::from(compiler);
        result
    }

    fn pass(&mut self, flag: &str) -> bool {
        self.pass.take(flag)
    }

    fn output(&mut self, flag: &str, it: &mut flags::FlagIterator) -> bool {
        if "-o" == flag {
            if let Some(output) = it.next() {
                self.output = Some(path::PathBuf::from(output));
            }
            return true;
        }
        false
    }

    fn flags(&mut self, flag: &str, it: &mut flags::FlagIterator) -> bool {
        //            # some parameters look like a filename, take those explicitly
        //            elif arg in {'-D', '-I'}:
        //                result.flags.extend([arg, next(args)])
        //            # and consider everything else as compile option.
        //            else:
        //                result.flags.append(arg)
        unimplemented!()
    }

    fn source(&mut self, file: &str) -> bool {
        //            # parameter which looks source file is taken...
        //            elif re.match(r'^[^-].+', arg) and classify_source(arg):
        //                result.files.append(arg)
        unimplemented!()
    }

    fn build(&self) -> Result<Vec<database::Entry>> {
        if !self.pass.is_compiling() {
            Err(ErrorKind::CompilationError("Compiler is not doing compilation.").into())
        } else if self.inputs.is_empty() {
            Err(ErrorKind::CompilationError("Have not found source files.").into())
        } else {
            let entries: Vec<_> = self
                .inputs
                .iter()
                .map(|input| {
                    let mut command: Vec<String> = Vec::new();
                    command.push(
                        self.compiler
                            .clone()
                            .into_os_string()
                            .into_string()
                            .unwrap(),
                    );
                    command.push("-c".to_string());
                    command.extend_from_slice(self.flags.as_ref());
                    command.push(input.clone().into_os_string().into_string().unwrap());
                    database::Entry {
                        directory: self.directory.clone(),
                        file: input.clone(),
                        command: command,
                        output: self.output.clone(),
                    }
                }).collect();
            Ok(entries)
        }
    }
}

/// Returns a value when the command is a compilation, None otherwise.
///
/// # Arguments
/// `classifier` - helper object to detect compiler
/// `command` - the command to classify
fn parse_command(
    classifier: &compiler::Classifier,
    working_dir: &str,
    command: &[String],
) -> Result<Vec<database::Entry>> {
    debug!("input was: {:?}", command);
    match classifier.split(command) {
        Some(compiler_and_parameters) => {
            let mut result =
                CompilerExecution::new(working_dir, compiler_and_parameters.0.as_str());
            let mut it = flags::FlagIterator::from(compiler_and_parameters.1);
            while let Some(arg) = it.next() {
                // if it's a pass modifier flag, update it and move on.
                if result.pass(arg.as_str()) {
                    continue;
                }
                // take the output flag
                if result.output(arg.as_str(), &mut it) {
                    continue;
                }
                // take flags
                if result.flags(arg.as_str(), &mut it) {
                    continue;
                }
                // take the rest as source file
                if result.source(arg.as_str()) {
                    continue;
                }
            }
            result.build()
        }
        _ => Err(ErrorKind::CompilationError("Compiler not recognized.").into()),
    }
}
